packages/overlay/src/Overlay.ts

```diff
@@ -0,0 +1,156 @@
+    type?: 'modal' | 'page' | 'hint' | 'auto' | 'manual';
+};
+
+/**
+ * @element sp-overlay
+ *
+ * @fires sp-opened - announces that an overlay has completed any entry animations
+ * @fires sp-closed - announce that an overlay has compelted any exit animations
+ */
+export class Overlay extends OverlayFeatures implements OverlayBase {
+    public static update(): void {
+        const overlayUpdateEvent = new CustomEvent('sp-update-overlays', {
+            bubbles: true,
+            composed: true,
+            cancelable: true,
+        });
+        document.dispatchEvent(overlayUpdateEvent);
+    }
+
+    public static async open(
+        target: HTMLElement,
+        interaction: TriggerInteractionsV1,
+        content: HTMLElement,
+        options: OverlayOptionsV1
+    ): Promise<() => void>;
+    public static async open(
+        content: HTMLElement,
+        options?: OverlayOptions
+    ): Promise<Overlay>;
+    public static async open(
+        targetOrContent: HTMLElement,
+        interactionOrOptions:
+            | TriggerInteractionsV1
+            | OverlayOptions
+            | undefined,
+        content?: HTMLElement,
+        options?: OverlayOptionsV1
+    ): Promise<Overlay | (() => void)> {
+        const v2 = arguments.length === 2;
+        const overlayContent = content || targetOrContent;
+        const overlay = new Overlay();
+        overlay.dispose = () => {
+            overlay.addEventListener('sp-closed', () => {
+                if (!restored) {
+                    restoreContent();
+                    restored = true;
+                }
+                requestAnimationFrame(() => {
+                    overlay.remove();
+                });
+            });
+            overlay.open = false;
+            overlay.dispose = noop;
+        };
+        let restored = false;
+        const restoreContent = reparentChildren([overlayContent], overlay, {
+            position: 'beforeend',
+            prepareCallback: (el) => {
+                const slot = el.slot;
+                el.removeAttribute('slot');
+                return () => {
+                    el.slot = slot;
+                };
+            },
+        });
+        if (v2) {
+            const options = interactionOrOptions as OverlayOptions;
+            overlay.append(overlayContent);
+            overlay.receivesFocus = options.receivesFocus ?? 'auto';
+            overlay.triggerElement = options.trigger || null;
+            overlay.type = options.type || 'modal';
+            overlay.offset = options.offset || 6;
+            overlay.placement = options.placement;
+            overlay.willPreventClose = !!options.notImmediatelyClosable;
+            requestAnimationFrame(() => {
+                requestAnimationFrame(() => {
+                    // Do we want to "open" this path, or leave that to the consumer?
+                    overlay.open = true;
+                });
+            });
```

Comment:
Are there guarantees we've made so we can rely on 2 frames always being perfect timing for this?

Is there an event we could listen for instead of timing this out by frame?

Reply:
This makes the assumption that the API consumer is immediately attaching the element to the DOM, so as to complete the `.open()` workflow. Two rAFs is the required timing to ensure that the append has occurred and the initial styling has been painted so that the open animation can occur as expected.

Possible alternative:
- `overlay.updateComplete.then(...)`, it _is_ a LitElement after all. not sure this fully accounts for the _next frame_ of the double rAF approach
- not opening the overlay and letting the consumer do it
- thoughts?

Reply:
If tests pass with it, I'd go for `overlay.updateComplete.then(...)`.

If that causes tests to fail, then we could put a pin in this for now. Using RAFs feels brittle, but other than your example of not appending immediately, I can't prove that this would break in any way.